exercises/react/src/example/kotlin/React.kt

```diff
@@ -0,0 +1,72 @@
+class Reactor<T>() {
+    abstract inner class Cell {
+        abstract val value: T
+        internal val dependents = mutableListOf<ComputeCell>()
+        internal fun eachDependent(f: (ComputeCell) -> Any) {
+            for (dep in dependents) {
+                f(dep)
+            }
+        }
+    }
+
+    interface Subscription {
+        fun cancel()
+    }
+
+    inner class InputCell(initialValue: T) : Cell() {
+        override var value: T = initialValue
+            set(newValue) {
+                field = newValue
+                eachDependent { it.propagate() }
+                eachDependent { it.fireCallbacks() }
+            }
+    }
+
+    inner class ComputeCell private constructor(val newValue: () -> T) : Cell() {
+        override var value: T = newValue()
+            private set
+
+        private var lastCallbackValue = value
+        private var callbacksIssued = 0
+        private val activeCallbacks = mutableMapOf<Int, (T) -> Any>()
+
+        constructor(c: Cell, f: (T) -> T) : this({ f(c.value) }) {
+            c.dependents.add(this)
+        }
+        constructor(c1: Cell, c2: Cell, f: (T, T) -> T) : this({ f(c1.value, c2.value) }) {
```

Comment:
OK, I have prepared what that looks like in the last commit, and we can decide whether it is better or worse to have it. Only disadvantage is as I mentioned the possibility of a caller mismatching the arities, giving one cell but expecting two values in the lambda, vice versa. Of course, since the test suite is in our control we know we haven't made that mistake, so the concern is only "what would happen if we used this as a library and gave it to other people to use, what kinds of mistakes might they make with it"

I personally like the varargs way slightly better.